File: scripts/ci/validate_frontmatter.py
#!/usr/bin/env python3
import sys
from pathlib import Path

def validate_frontmatter(root: Path, exclude_prefixes=('.kanban', 'node_modules', '.git')):
    files = [
        p for p in root.rglob('*.md')
        if not any(str(p.relative_to(root)).startswith(pre) for pre in exclude_prefixes)
    ]
    failures = []
    for path in files:
        text = path.read_text()
        if not text.startswith('---'):
            failures.append((str(path), 'missing frontmatter block'))
            continue
        end = text.find('---', 3)
        if end == -1:
            failures.append((str(path), 'unclosed frontmatter'))
            continue
        fm = text[3:end]
        fm_lines = fm.strip().split('\n')
        fm_keys = {}
        curr_key = None
        for line in fm_lines:
            if ':' in line and not line.startswith(' ') and not line.startswith('\t'):
                k, v = line.split(':', 1)
                curr_key = k.strip()
                fm_keys[curr_key] = v.strip()
            elif curr_key:
                fm_keys[curr_key] += ' ' + line.strip()

        for key in ['title', 'type', 'status', 'description']:
            if key not in fm_keys:
                failures.append((str(path), f'missing {key}'))
            elif not fm_keys[key] or fm_keys[key] in ('>', '|', "''", '""'):
                failures.append((str(path), f'empty {key} value'))
    if failures:
        print('FRONTMATTER_FAILURES')
        for path, reason in failures:
            print(f'  {path}: {reason}')
        return 1
    print(f'FRONTMATTER_OK checked={len(files)}')
    return 0

if __name__ == '__main__':
    root = Path(sys.argv[1]) if len(sys.argv) > 1 else Path('.')
    sys.exit(validate_frontmatter(root))

File: scripts/ci/validate-frontmatter.ts
import fs from 'fs';
import path from 'path';

async function getFiles(dir: string, files: string[] = [], excludePrefixes: string[] = ['.kanban', 'node_modules']): Promise<string[]> {
    try {
        const fileList = await fs.promises.readdir(dir);
        for (const file of fileList) {
            const fullPath = path.join(dir, file);
            const relPath = path.relative(process.cwd(), fullPath);
            if (excludePrefixes.some(pre => relPath.startsWith(pre) || relPath.includes('/' + pre + '/'))) {
                continue;
            }
            const stat = await fs.promises.stat(fullPath);
            if (stat.isDirectory()) {
                await getFiles(fullPath, files, excludePrefixes);
            } else if (fullPath.endsWith('.md')) {
                files.push(fullPath);
            }
        }
    } catch(e) {}
    return files;
}

async function validateFrontmatter(rootDir: string): Promise<number> {
    const root = path.resolve(process.cwd(), rootDir);
    const files = await getFiles(root);
    const failures: { file: string; reason: string }[] = [];

    for (const filePath of files) {
        const text = await fs.promises.readFile(filePath, 'utf8');
        if (!text.startsWith('---')) {
            failures.push({ file: filePath, reason: 'missing frontmatter block' });
            continue;
        }
        const end = text.indexOf('---', 3);
        if (end === -1) {
            failures.push({ file: filePath, reason: 'unclosed frontmatter' });
            continue;
        }
        const fm = text.substring(3, end);
        for (const key of ['title', 'type', 'status', 'description']) {
            if (!fm.includes(`${key}:`)) {
                failures.push({ file: filePath, reason: `missing ${key}` });
            }
        }
    }

    if (failures.length > 0) {
        console.log('FRONTMATTER_FAILURES');
        for (const failure of failures) {
            console.log(`  ${failure.file}: ${failure.reason}`);
        }
        return 1;
    }
    console.log(`FRONTMATTER_OK checked=${files.length}`);
    return 0;
}

const rootArg = process.argv[2] || '.';
validateFrontmatter(rootArg).then(exitCode => process.exit(exitCode));
